### DISCUSSION TITLE
NextAuth session callback does not trigger in production

### DISCUSSION THREAD

**Discussion by BlindBMan** [nextauthjs/next-auth#5988] [Help]
I deployed a `Nextjs`(v13) app with AWS Amplify and using `NextAuth`(v4.17.0). I'm using `CredentialsProvider` with a custom server. All works great in development environment, but in production the `session` callback doesn't fire and the session is empty, even if the token gets created in the database

`/page/api/auth/[...nextauth].tsx` _disregard the console logs lol_

```js
import NextAuth from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import jwt_decode from "jwt-decode";
import { TokenInfo } from "../../../components/types/auth_types";


async function refreshAccessToken(token) {
    try {
        console.log("BUT WHY?");

        const res = await fetch(
            `${process.env.NEXT_PUBLIC_API_URL}/api/token/refresh/`,
            {
                method: "POST",
                body: JSON.stringify({refresh: token.refreshToken}),
                headers: {"Content-Type": "application/json"},
            }
        );

        if (!res.ok) throw "refreshError";
        const responseJson = await res.json();
        return {
            ...token,
            accessToken: responseJson.access,
        }

    } catch(error) {
        return {
            ...token,
            error: "RefreshAccessTokenError",
        }
    }
}

export const authOptions = {
    providers: [
        CredentialsProvider({
            id: "credentials",
            name: "Credentials",
            credentials: {
                email: { label: "Username", type: "text", placeholder: "" },
                password: { label: "Password", type: "password" }
            },
            async authorize(credentials, req) {
                const userCredentials = {
                    email: credentials.email, password: credentials.password
                };
                try {
                    const res = await fetch(
                        `${process.env.NEXT_PUBLIC_API_URL}/api/token/`,
                        {
                            method: "POST",
                            body: JSON.stringify(userCredentials),
                            headers: {"Content-Type": "application/json"},
                            credentials: "include",
                        }
                    );
                    console.log("res", res);
                    if (res.ok) {
                        const responseJson = await res.json();
                        console.log("resJson", responseJson);
                        const tokenInfo: TokenInfo = jwt_decode(responseJson.access);
                        console.log("tokenInfo", tokenInfo);
                        return {
                            id: tokenInfo.user_id.toString(),
                            email: tokenInfo.email,
                            firstName: tokenInfo.first_name,
                            lastName: tokenInfo.last_name,
                            isStaff: tokenInfo.is_staff,
                            accessToken: responseJson.access,
                            refreshToken: responseJson.refresh,
                        };
                    }
                    return null;
                } catch(e) {
                    return null;
                }
            }
        })
    ],
    callbacks: {
        async jwt({ token, account, user }) {
            if (account && user) {
                console.log("got into token", user);
                
                token.firstName = user.firstName;
                token.lastName = user.lastName;
                token.refreshToken = user.refreshToken;
                token.accessToken = user.accessToken;
            }
            if (token.accessToken) {
                console.log("got in this if instead")
                const decodedToken: TokenInfo = jwt_decode(token.accessToken);
                if (Date.now() < decodedToken.exp * 1000) {
                    console.log("got here, returned properly");
                    return token;
                }
            }
            console.log("got here, not properly, why?");
            return await refreshAccessToken(token);
        },
        async session({ session, token }) {
            console.log("getting session");
            session.user.firstName = token.firstName;
            session.user.lastName = token.lastName;
            session.accessToken = token.accessToken;
            console.log("sess", session);
            return session;
        }
    },
    secret: process.env.NEXT_AUTH_SECRET,
    session: {
        maxAge: 2 * 24 * 60 * 60,  // two days
    }
};

export default NextAuth(authOptions);
```

I have searched as best as I could and couldn't find anything that I didn't do already.

My understanding is I don't need to set `session: { strategy: "jwt"}` since that's the default.

I have `NEXT_AUTH_SECRET="mysecret"`, `NEXT_PUBLIC_API_URL="https://www.backend_domain.com"` and `NEXTAUTH_URL="https://www.frontend_domain.com"` set properly in `.env.production` and the API calls succeed, as well as the `NextAuth` calls return `200` status code, no errors in Amplify logs

Let me know if I can provide more info

Edit 1:

If I navigate to `/api/auth/signin/credentials` and use the default UI for login, the session gets created successfully

**Comment by FaresKi** [comment]
Have you managed to find a solution, despite it being a year? lol

**Reply by BlindBMan** [reply]
I was unable to pinpoint what exactly caused this or reproduce it after it apparently fixed itself. I did not make any changes to the `nextjs` code or AWS, as far as I remember

**Comment by hect1c** [comment]
Interesting, I've recently got the same issue deploying to serverless environment. It was working before but then I updated the package I use to build and deploy (SST https://sst.dev/) my app is when I noticed this issue. 

Upoon I login I can see the jwt callback has the updated object from my backend response but then the session callback does not get triggered and I can't figure out why or what's causing it. Trying to debug but struggling to see at which point this issue takes place. Also in the SST community to see if they can help.

**Comment by DevinLeamy** [comment]
Based on my debugging, next-auth does not seem to use the callbacks if a JWT has already been issued (i.e. you already have a session and are logged in). Therefore, if you made a change to your callbacks you will need to call `signOut` (from `"next-auth/react"`) before you'll see those changes. 


TL;DR try logging out and logging back in.
